File: ref/src/calculations/marketSummary.ts
import type { CalculationOutput, MarketSignal, MarketSummary } from "../types/calculations";
import type { Candle, SymbolCode, Timeframe } from "../types/market";

export function summarizeMarket(
  symbol: SymbolCode,
  timeframe: Timeframe,
  candles: Candle[],
  calculationOutputs: Record<string, CalculationOutput> = {}
): MarketSummary {
  if (candles.length < 2) {
    const latest = candles[candles.length - 1];
    const latestPrice = latest?.close ?? 0;
    return {
      symbol,
      timeframe,
      latestPrice,
      latestTimestamp: latest?.timestamp ?? "",
      changePercentFromFirstVisible: 0,
      visibleHigh: latest?.high ?? latestPrice,
      visibleLow: latest?.low ?? latestPrice,
      visibleVolume: latest?.volume ?? 0,
      averageVolume: latest?.volume ?? 0,
      realizedVolatility: 0,
      trend: "insufficient_data",
      notableSignals: []
    };
  }

  const first = candles[0];
  const latest = candles[candles.length - 1];
  const visibleHigh = Math.max(...candles.map((candle) => candle.high));
  const visibleLow = Math.min(...candles.map((candle) => candle.low));
  const visibleVolume = candles.reduce((total, candle) => total + candle.volume, 0);
  const averageVolume = visibleVolume / candles.length;
  const changePercentFromFirstVisible = ((latest.close - first.close) / first.close) * 100;
  const returns = candles.slice(1).map((candle, index) => Math.log(candle.close / candles[index].close));
  const realizedVolatility = Math.sqrt(returns.reduce((total, value) => total + value ** 2, 0) / returns.length) * 100;
  const trend = classifyTrend(changePercentFromFirstVisible);
  const notableSignals = collectSignals(symbol, candles, averageVolume, visibleHigh, visibleLow, calculationOutputs);

  return {
    symbol,
    timeframe,
    latestPrice: latest.close,
    latestTimestamp: latest.timestamp,
    changePercentFromFirstVisible: round(changePercentFromFirstVisible),
    visibleChangeBaseTimestamp: first.timestamp,
    visibleChangeBaseClose: first.close,
    visibleHigh,
    visibleLow,
    visibleVolume,
    averageVolume,
    realizedVolatility: round(realizedVolatility),
    trend,
    notableSignals
  };
}

function classifyTrend(changePercent: number): MarketSummary["trend"] {
  if (changePercent >= 2.5) return "strong_up";
  if (changePercent >= 0.45) return "up";
  if (changePercent <= -2.5) return "strong_down";
  if (changePercent <= -0.45) return "down";
  return "sideways";
}

function collectSignals(
  symbol: SymbolCode,
  candles: Candle[],
  averageVolume: number,
  visibleHigh: number,
  visibleLow: number,
  calculationOutputs: Record<string, CalculationOutput>
): MarketSignal[] {
  const latest = candles[candles.length - 1];
  const signals: MarketSignal[] = [];
  if (latest.volume > averageVolume * 1.8) {
    signals.push({
      type: "volume_spike",
      severity: latest.volume > averageVolume * 2.5 ? "high" : "medium",
      message: `${symbol} volume is elevated versus the visible average.`,
      timestamp: latest.timestamp
    });
  }
  if (latest.high >= visibleHigh) {
    signals.push({
      type: "new_visible_high",
      severity: "medium",
      message: `${symbol} is pressing the visible-range high.`,
      timestamp: latest.timestamp
    });
  }
  if (latest.low <= visibleLow) {
    signals.push({
      type: "new_visible_low",
      severity: "medium",
      message: `${symbol} is testing the visible-range low.`,
      timestamp: latest.timestamp
    });
  }

  for (const output of Object.values(calculationOutputs)) {
    const rsi = output.series.find((series) => series.key === "rsi");
    const lastRsi = rsi?.points[rsi.points.length - 1]?.value;
    if (typeof lastRsi === "number" && lastRsi >= 70) {
      signals.push({
        type: "rsi_overbought",
        severity: "medium",
        message: `${symbol} RSI is above 70.`,
        timestamp: latest.timestamp
      });
    }
    if (typeof lastRsi === "number" && lastRsi <= 30) {
      signals.push({
        type: "rsi_oversold",
        severity: "medium",
        message: `${symbol} RSI is below 30.`,
        timestamp: latest.timestamp
      });
    }
  }

  return signals.slice(0, 6);
}

function round(value: number): number {
  return Number(value.toFixed(4));
}

File: ref/backend/market_summary.py
from __future__ import annotations

import math

from backend.schemas import Candle, MarketSignal, MarketSummary


def calculate_market_summary(symbol: str, timeframe: str, candles: list[Candle]) -> MarketSummary:
    if len(candles) < 2:
        latest = candles[-1] if candles else None
        timestamp = latest.timestamp if latest else ""
        price = latest.close if latest else 0.0
        return MarketSummary(
            symbol=symbol,
            timeframe=timeframe,  # type: ignore[arg-type]
            latestPrice=price,
            latestTimestamp=timestamp,
            changePercentFromFirstVisible=0.0,
            visibleHigh=price,
            visibleLow=price,
            visibleVolume=latest.volume if latest else 0.0,
            averageVolume=latest.volume if latest else 0.0,
            realizedVolatility=0.0,
            trend="insufficient_data",
            notableSignals=[],
        )

    first = candles[0]
    latest = candles[-1]
    visible_high = max(candle.high for candle in candles)
    visible_low = min(candle.low for candle in candles)
    visible_volume = sum(candle.volume for candle in candles)
    average_volume = visible_volume / len(candles)
    change_percent = ((latest.close - first.close) / first.close) * 100 if first.close else 0.0
    returns = []
    for previous, current in zip(candles, candles[1:]):
        if previous.close > 0 and current.close > 0:
            returns.append(math.log(current.close / previous.close))
    realized_volatility = math.sqrt(sum(value * value for value in returns) / len(returns)) * 100 if returns else 0.0

    if change_percent >= 2.5:
        trend = "strong_up"
    elif change_percent >= 0.45:
        trend = "up"
    elif change_percent <= -2.5:
        trend = "strong_down"
    elif change_percent <= -0.45:
        trend = "down"
    else:
        trend = "sideways"

    signals: list[MarketSignal] = []
    if latest.volume > average_volume * 1.8:
        signals.append(
            MarketSignal(
                type="volume_spike",
                severity="high" if latest.volume > average_volume * 2.5 else "medium",
                message=f"{symbol} volume is elevated versus the visible average.",
                timestamp=latest.timestamp,
            )
        )
    if latest.high >= visible_high:
        signals.append(
            MarketSignal(
                type="new_visible_high",
                severity="medium",
                message=f"{symbol} is pressing the visible-range high.",
                timestamp=latest.timestamp,
            )
        )
    if latest.low <= visible_low:
        signals.append(
            MarketSignal(
                type="new_visible_low",
                severity="medium",
                message=f"{symbol} is testing the visible-range low.",
                timestamp=latest.timestamp,
            )
        )

    return MarketSummary(
        symbol=symbol,
        timeframe=timeframe,  # type: ignore[arg-type]
        latestPrice=latest.close,
        latestTimestamp=latest.timestamp,
        changePercentFromFirstVisible=round(change_percent, 4),
        visibleHigh=visible_high,
        visibleLow=visible_low,
        visibleVolume=visible_volume,
        averageVolume=average_volume,
        realizedVolatility=round(realized_volatility, 4),
        trend=trend,  # type: ignore[arg-type]
        notableSignals=signals[:5],
    )
